cloud/util/util.go

```diff
@@ -0,0 +1,133 @@
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	infrastructurev1beta1 "github.com/oracle/cluster-api-provider-oci/api/v1beta1"
+	"github.com/oracle/cluster-api-provider-oci/cloud/config"
+	"github.com/oracle/cluster-api-provider-oci/cloud/scope"
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetClusterIdentityFromRef returns the AzureClusterIdentity referenced by the OCICluster.
+func GetClusterIdentityFromRef(ctx context.Context, c client.Client, ociClusterNamespace string, ref *corev1.ObjectReference) (*infrastructurev1beta1.OCIClusterIdentity, error) {
+	identity := &infrastructurev1beta1.OCIClusterIdentity{}
+	if ref != nil {
+		namespace := ref.Namespace
+		if namespace == "" {
+			namespace = ociClusterNamespace
+		}
+		key := client.ObjectKey{Name: ref.Name, Namespace: namespace}
+		if err := c.Get(ctx, key, identity); err != nil {
+			return nil, err
+		}
+		return identity, nil
+	}
+	return nil, nil
+}
+
+// GetOrBuildClientFromIdentity creates ClientProvider from OCIClusterIdentity object
+func GetOrBuildClientFromIdentity(ctx context.Context, c client.Client, identity *infrastructurev1beta1.OCIClusterIdentity, regionOverride string) (*scope.ClientProvider, error) {
+	if identity.Spec.Type == infrastructurev1beta1.UserPrincipal {
+		secretRef := identity.Spec.PrincipalSecret
+		key := types.NamespacedName{
+			Namespace: secretRef.Namespace,
+			Name:      secretRef.Name,
+		}
+		secret := &corev1.Secret{}
+
+		if err := c.Get(ctx, key, secret); err != nil {
+			return nil, errors.Wrap(err, "Unable to fetch ClientSecret")
+		}
+
+		tenancyId := string(secret.Data[config.Tenancy])
+		userId := string(secret.Data[config.User])
+		fingerPrint := string(secret.Data[config.Fingerprint])
+		passphrase := string(secret.Data[config.Passphrase])
+		privatekey := string(secret.Data[config.Key])
+		region := string(secret.Data[config.Region])
+		// set the default region if not provided in the secret
+		if region == "" {
+			region = regionOverride
+		}
+		conf := common.NewRawConfigurationProvider(
+			tenancyId,
+			userId,
+			region,
+			fingerPrint,
+			privatekey,
+			common.String(passphrase))
+
+		clientProvider, err := scope.NewClientProvider(conf)
+		if err != nil {
+			return nil, err
+		}
+		return clientProvider, nil
+	}
+	return nil, errors.New(fmt.Sprintf("invalid oci principal format type: %s", identity.Spec.Type))
+}
+
+// IsClusterNamespaceAllowed indicates if the cluster namespace is allowed.
+func IsClusterNamespaceAllowed(ctx context.Context, k8sClient client.Client, allowedNamespaces *infrastructurev1beta1.AllowedNamespaces, namespace string) bool {
+	if allowedNamespaces == nil {
+		return false
+	}
+
+	// empty value matches with all namespaces
+	if reflect.DeepEqual(*allowedNamespaces, infrastructurev1beta1.AllowedNamespaces{}) {
```

Comment:
`allowedNamespaces` is nullable correct? will this cause a dereference error?

Reply:
there is a nil check in line 95